src/actions/firebase/firebase.ts

```diff
@@ -0,0 +1,167 @@
+      }
+      return new Hub.ActionResponse(response)
+	  }
+
+    async verifyAndSendMessage(params: Hub.ParamMap, imageName: string): Promise<any> {
+      return new Promise(async (resolve, reject) => {
+        let data: any = params.data
+        if (!params.title) {
+          reject("Needs a valid title.")
+        }
+        if (!params.deviceIds) {
+          resolve()
+        }
+        let notification: any = {title: params.title}
+        let notificationData: any = data
+        notificationData.imageUrl = imageName
+        if (params.timeToLive) {
+          this.notificationOptions.timeToLive = parseFloat(params.timeToLive)
+        }
+        if (params.priority) {
+          this.notificationOptions.priority = params.priority
+        }
+        try {
+          const userObj = JSON.parse(params.deviceIds ?? "[]") as any[]
+          let deviceIdObject = params.deviceIds as unknown as any[]
+          if (deviceIdObject.length == 0) {
+            resolve()
+          }
+          await userObj.forEach(async (userDevices) => {
+            let devices = userDevices as any[]
+            await devices.forEach(async (device, index) => {
+              if (device.device_id && device.user_id) {
+                let deviceId = device.device_id.toString()
+                notificationData.userId = device.user_id.toString()
+                let payload = {
+                  notification: notification,
+                  data: notificationData
+                };
+                try {
+                  await this.sendMessageToDevice(deviceId, payload, this.notificationOptions)
+                }
+                catch (error) {
+                  reject(error)
+                }
+              }
+              if (index === devices.length -1) {
+                resolve()
+              }
+            })
+          });
+        } catch (error) {
+          reject(error)
+        }
+      })
+      .catch((error) => {
+        throw error
+      })
+    }
+
+    async sendMessageToDevice(deviceId: string, payload: admin.messaging.MessagingPayload, options?: admin.messaging.MessagingOptions): Promise<any> {
+      return new Promise((resolve, reject) => {
+        FirebaseAction.setFirebaseClient()
+        admin.messaging().sendToDevice(deviceId, payload, options)
+          .then( (_: any) => {
+            resolve()
+          })
+          .catch( (error: any) => {
+            reject(error.message)
+          });
+      })
+      .catch((error) => {
+        throw error
+      });
+    }
+
+    async uploadImage(request: Hub.ActionRequest): Promise<any> {
+      let data = request.formParams.data as any
+
+      try {
+        if (request.attachment?.dataBuffer) {
+          return new Promise((resolve, reject) => {
+            FirebaseAction.setFirebaseClient()
+            let dateString = new Date().toISOString()
+            let alertId = data.alertId ?? ""
+            const destFileName = dateString + alertId;
```

Comment:
alert id can be non unique across looker instances so may be we need something more to make this unique..